lib/httpception.js

```diff
@@ -79,69 +60,44 @@ module.exports = function httpception(mockDefinition, promiseFactory) {
-                        resolveFromAfterEach = resolve;
-                    }).then(() => {
-                        // At this point we know that afterEach is ready for us
-                        // to handle the response, but the mitmAssertion might
-                        // already have rejected the promise. If `deferredError`
-                        // is NOT undefined, we need to reject the outer
-                        // promise.
-                        //   If mitmAssertion rejected early, we had no where to
-                        // pass the rejection to, which is why we need this
-                        // nested promise trickery.
-                        if (deferredError) {
-                            outerReject(deferredError);
-                        }
-                    });
+    if (promiseFactory) {
+        resolveFromAfterEach = false;
 
-                    return innerPromise;
-                },
-                'with http mocked out allowing modification',
-                mockDefinitionForTheCurrentTest,
-                'not to error'
-            );
+        return mocker.mock(promiseFactory).then(function () {
+            return expect(promiseFactory, 'to perform HTTP traffic', mocker.descriptions);
+        }).then(function () {
+            mocker = undefined;
+        }).catch(function (e) {
+            mocker = undefined;
+            throw e;
+        });
+    } else if (!promise) {
+        var resolvePromise = expect.promise(function (resolve, reject) {
+            resolveFromAfterEach = resolve;
+        });
 
-            mitmAssertion.then(() => {
-                // mitmAssertion will only resolve if there's no errors in the
-                // mocked out traffic, which means that at this point we know
-                // that the innerPromise have run, which will only happen after
-                // afterEach has run, so we can safely resolve the promise, as
-                // we know afterEach will be waiting for it.
-                outerResolve();
-            }, err => {
-                if (!resolveFromAfterEach) {
-                    // If resolveFromAfterEach has been unset, we know that
-                    // afterEach is waiting for the response already. Rejecting
-                    // the outer promise immediately and not setting the err to
-                    // avoid the innerPromise trying to reject again.
-                    return outerReject(err);
-                }
-                // Defer rejection until innerPromise has been resolved.
-                deferredError = err;
+        var mockerPromise = expect.promise(function () {
+            return mocker.mock(function () {
+                return resolvePromise;
             });
         });
+
+        promise = Promise.all([
+            resolvePromise,
+            mockerPromise
+        ]).then((result) => {
+            return expect(function afterEach() {}, 'to perform HTTP traffic', mocker.descriptions);
```

Comment:
This looks a bit weird with the empty function? It suggests that the assertion should be shuffled around a bit so it can be something like `expect(mocker, 'to have performed the HTTP traffic that it was supposed to')`? :)

Reply:
Ahhh yeah - the reason it looks this way is actually to trick Unexpected into inspecting it (so you get a hint it was in afterEach) but then to output the same assertion as in the promise factory case - having `httpception` output two different assertions felt like an output inconsistency.

Reply:
The only other way I could think of doing it was to get really deep into overriding the error message and I wanted to avoid that as part of this cleanup..

Reply:
Fair :)